cpp/src/unary/math_ops.cu

```diff
@@ -198,80 +204,127 @@ private:
 
 public:
     template <typename T>
-    typename std::enable_if_t<is_supported<T>(), gdf_error>
-    operator()(gdf_column *input, gdf_column *output) {
-        return cudf::unary::Launcher<T, cudf::bool8, F>::launch(input, output);
+    typename std::enable_if_t<is_supported<T>(), void>
+    operator()(gdf_column const* input, gdf_column *output) {
+        cudf::unary::Launcher<T, cudf::bool8, F>::launch(input, output);
     }
 
     template <typename T>
-    typename std::enable_if_t<!is_supported<T>(), gdf_error>
-    operator()(gdf_column *input, gdf_column *output) {
-        return GDF_UNSUPPORTED_DTYPE;
+    typename std::enable_if_t<!is_supported<T>(), void>
+    operator()(gdf_column const* input, gdf_column *output) {
+        CUDF_FAIL("Unsupported datatype for operation");
     }
 };
 
+} // namespace detail
+
+gdf_column unary_operation(gdf_column const& input, unary_op op) {
+
+    gdf_column output;
+
+    if (op == unary_op::NOT)
+    {
+        // TODO: replace this with a proper column constructor once
+        // cudf::column is implemented
+        bool allocate_mask = (input.valid != nullptr);
+        output = cudf::allocate_column(GDF_BOOL8, input.size, allocate_mask);
+    }
+    else
+        output = cudf::allocate_like(input);
+
+    if (input.size == 0) return output;
```

Comment:
This should be moved above such that we return an empty column of the correct type when the input is empty.

Reply:
Actually `cudf::allocate_like` and `cudf::allocate_column` will not allocate anything if `input.size == 0`. But by keeping this line here, we can ensure that the result of `NOT` is an empty column of type `GDF_BOOL8`.

Reply:
Yeah, that's a good point.